tests/js/plugin/enhancedanchoring-test.coffee

```diff
@@ -0,0 +1,205 @@
+      code: (annotation, target) ->
+        new FakeAnchor am,annotation, target
+
+    am
+
+  addDummyData = ->
+    target11 =
+      id: "target 1.1"
+    target12 =
+      id: "target 1.2"
+    target2 =
+      id: "target 2"
+    target3 =
+      id: "target 3"
+    ann1 =
+      id: "annotation 1"
+      anchors: []
+    ann2 =
+      id: "annotation 2"
+      anchors: []
+    ann3 =
+      id: "annotation 3"
+      anchors: []
+
+    ann1.target = [target11, target12]
+    ann2.target = [target2]
+    ann3.target = [target3]
+
+    anchor11 = am.createAnchor(ann1, target11).result
+    anchor12 = am.createAnchor(ann1, target12).result
+    anchor2 = am.createAnchor(ann2, target2).result
+    anchor3 = am.createAnchor(ann3, target3).result
+
+    hl11 = anchor11.highlight[0]
+    hl12 = anchor12.highlight[0]
+    hl2 = anchor2.highlight[0]
+    hl3 = anchor3.highlight[0]
+
+  describe "Basics", ->
+
+    beforeEach ->
+      initAnchoring()
+      addDummyData()
+
+    it 'can be initialized', ->
+      assert.ok(am)
+
+    it 'it can do some anchoring', ->
+      assert.equal anchor11.id, "fake anchor for target 1.1"
+
+    it 'Annotations have references to their anchors', ->
+      assert.isArray ann1.anchors
+      assert.include ann1.anchors, anchor11
+      assert.include ann1.anchors, anchor12
+      assert.equal ann1.anchors.length, 2
+
+      assert.isArray ann2.anchors
+      assert.include ann2.anchors, anchor2
+      assert.equal ann2.anchors.length, 1
+
+      assert.isArray ann3.anchors
+      assert.include ann3.anchors, anchor3
+      assert.equal ann3.anchors.length, 1
+
+    it 'Anchors have references to the annotations', ->
+      assert.equal anchor11.annotation, ann1
+      assert.equal anchor12.annotation, ann1
+      assert.equal anchor2.annotation, ann2
+      assert.equal anchor3.annotation, ann3
+
+    it 'Anchors have references to the targets', ->
+      assert.equal anchor11.target, target11
+      assert.equal anchor12.target, target12
+      assert.equal anchor2.target, target2
+      assert.equal anchor3.target, target3
+
+    it 'Anchors can also be found by the page', ->
+      assert.include am.anchors[anchor11.startPage], anchor11
+      assert.include am.anchors[anchor12.startPage], anchor12
+      assert.include am.anchors[anchor2.startPage], anchor2
+      assert.include am.anchors[anchor3.startPage], anchor3
+
+    it 'Anchors have links to the highlights per page', ->
+      assert.isObject anchor11.highlight
+      assert.ok anchor11.highlight[0]
+      assert.include anchor11.highlight[0], hl11
+      assert.equal hl11.page, 0
+
+    it 'Highlights have links to the anchor', ->
+      assert.equal hl11.anchor, anchor11
+
+  describe "getAnchors", ->
+
+    it 'returns an empty array by default', ->
+      initAnchoring()
+      anchors = am.getAnchors()
+      assert.isArray anchors
+      assert.equal anchors.length, 0
+
+    beforeEach ->
+      initAnchoring()
+      addDummyData()
```

Comment:
On 2015-02-05 16:55, Aron Carroll wrote:
> Sorry to push this, but it does run before all three tests. The reason
> that the tests still pass is that the first test calls |initAnchoring|
> again which blows away the one defined in the |beforeEach| block.

Oh I see. Yes, you are right.
(However, this no longer matters, since I moved the forEach block to the
top, as suggested.)